sherpa/ui/utils.py

```diff
@@ -5186,8 +5409,38 @@ def notice_id(self, ids, lo=None, hi=None, **kwargs):
         if lo is not None and type(lo) in (str, numpy.string_):
             return self._notice_expr_id(ids, lo, **kwargs)
 
-        for i in ids:
-            self.get_data(i).notice(lo, hi, **kwargs)
+        # Unlike notice() we do not sort the id list as this
+        # was set by the user.
+        #
+        # The bkg_id argument is astro-specific. The need
+        # to use it to access the correct data object is
+        # problematic.
+        #
+        try:
+            bkg_id = kwargs["bkg_id"]
+            del kwargs["bkg_id"]  # important to remove
+        except KeyError:
+            bkg_id = None
+
+        for idval in ids:
+
+            idstr = f"dataset {idval}"
+            data = self.get_data(idval)
+            if bkg_id is not None:
+                data = data.get_background(bkg_id)
+                idstr += f": background {bkg_id}"
+
+            ofilter = _get_filter(data)
+            data.notice(lo, hi, **kwargs)
+            nfilter = _get_filter(data)
+
+            try:
+                xlabel = data.get_xlabel()
+            except AttributeError:
+                # Data2D case
+                xlabel = None
+
+            report_filter_change(idstr, ofilter, nfilter, xlabel)
```

Comment:
I'm a little worried that this is almost (or entirely?) identical to the implementation of `notice`. Is notice not simply a call do `notice_ids` with all datasets selected? If so, why do we copy the implementation? The more lines of code we need, to more it seems to me that we should try to unify. If it's only a few lines, OK, but this PR considerably lengthens the implementation as well as the docstring. 
Something like:
```python
def notice(self, lo=None, hi=None, **kwargs):
    self.notice_ids(self.list_Data_ids(), lo, hi, **kwargs)
```


For the docstring, can't we just have the full examples and description for one of them and say "See docs to notice_id for examples"?
If I were to implement that again, I would make just one function with `ids` as an optional parameter, if given it selects the dataset, if not, you get all datasets. That's too late now, but at least we can try to cut down an duplicated implementation.

Reply:
In #1586 I've noted that the 2D code needs re-working to remove duplication. We can extend that to here too, but I don't have time/energy/desire to make significant changes (even if they should be logically obviously correct, I've learnt not to trust this belief too much...) at this time.

Reply:
However, saying that, I can remove *some* repetition of code here.

Reply:
I've moved this code to the top-level in `notice_data_range`. We can look at more-extensive rework for the next release.